### PR TITLE
Add PlaceDefenseTowardsEnemyChance trait to BaseBuilderBotModule

### PR DESCRIPTION
**Pull request by blackhand1001**
Add PlaceDefenseTowardsEnemyChance trait to basebuilderbotmodule. This defaults to 100 which is the current behavior. This change now allows you to set the chance that bots will place defenses evenly around the base like the AI in stock red alert and Tiberian sun did.

### CODE REVIEW

**Comment by blackhand1001** [comment]
Here is a comparison of the two modes:

True:
OpenRA-2019-10-16T005255498Z

False:
OpenRA-2019-10-16T003934916Z

File: OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs]: It would be more flexible (and IMO something we'd definitely want to enable in the default mods) if we made this `int PlaceDefenseTowardsEnemyChance = 100` and changed the check below to a comparison against `world.LocalRandom.Next(100)`. This would mix random and directed placement.

Comment (GraionDilach) [reply OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs]: Except that this stacking-defenses "feature" is useless and broken beyond belief in it's current state. The defense placement ignores the terrain and pathfinding and in games with multiple players, ends up stacking towards one random base, leaving it open for attacks from any of the other directions.

In my local tests upon 6P FFA AI simulations on Defcon 6 (so this probably applies to Vegetation too, since their layout is very similar), the AI performed way worse with this shipped stacked approach (ultimately I locally removed this special behaviour from my AI module).

Comment (blackhand1001) [reply OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs]: Changed it into an integer with a probability.

**Comment by abcdefg30** [comment]
Do we want to use the new settings and tweak the AIs in this PR, or leave it to a follow-up?

File: OpenRA.Mods.Common/Traits/BotModules/BotModuleLogic/BaseBuilderQueueManager.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/BotModuleLogic/BaseBuilderQueueManager.cs]: I wonder if we want to rename this now.

**Comment by blackhand1001** [comment]
I'll do a seperate PR for the AI modules. I need to tweak the ratio of defenses up a bit higher since now they will be built all around the base.

File: OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs]: This needs a full stop at the end.

**Review by abcdefg30 — COMMENTED** [review]
Code changes look good to me after the one comment below is resolved. Please squash the commits together as well.

**Review by abcdefg30 — APPROVED** [review]

**Comment by pchote** [comment]
Looks reasonable at a quick look, but I'm not going to be able to test this any time soon.

File: OpenRA.Mods.Common/Traits/BotModules/BotModuleLogic/BaseBuilderQueueManager.cs
Comment (matjaeck) [reply OpenRA.Mods.Common/Traits/BotModules/BotModuleLogic/BaseBuilderQueueManager.cs]: I also find it a bit strange that the `PlaceDefenseTowardsEnemyChance` parameter dictates a `BuildingType`. We discussed it on IRC and pchote suggested renaming the enum to `PlacementType` or similar.

**Review by ghost — COMMENTED** [review]
Works as expected and LGTM after it's decided if `BuildingType` should be renamed or not.

**Review by ghost — APPROVED** [review]
Renaming can be done in the eventual refactor of the module.

**Comment by teinarss** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/4086ba5cb4b279ea2da2b66a591eff75f8d13f45...062bb2b5ddca51fd67c85f76b4792405aff958e1

**Pull request merged**
